File: src/09.tree/24.623.go
package tree

func AddOneRow1(root *TreeNode, val int, depth int) *TreeNode {
	if depth == 1 {
		return &TreeNode{val, root, nil}
	}

	var dfs func(*TreeNode, int)

	dfs = func(tn *TreeNode, i int) {
		if i+1 == depth {
			tn.Left = &TreeNode{val, tn.Left, nil}
			tn.Right = &TreeNode{val, nil, tn.Right}
			return
		}

		if tn.Left != nil {
			dfs(tn.Left, i+1)
		}

		if tn.Right != nil {
			dfs(tn.Right, i+1)
		}
	}

	dfs(root, 1)

	return root
}

func AddOneRow(root *TreeNode, val int, depth int) *TreeNode {
	if depth == 1 {
		return &TreeNode{val, root, nil}
	}

	queue := []*TreeNode{root}
	cur := 1
	for len(queue) > 0 {
		size := len(queue)

		for i := size; i > 0; i-- {
			node := queue[0]
			queue = queue[1:]

			if cur+1 == depth {
				node.Left = &TreeNode{val, node.Left, nil}
				node.Right = &TreeNode{val, nil, node.Right}
			}

			if node.Left != nil {
				queue = append(queue, node.Left)
			}

			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}

		cur++
		if cur == depth {
			return root
		}
	}

	return root
}

File: src/09.tree/24.623.ts
import { TreeNode } from "./01.100";

export function addOneRow1(
    root: TreeNode | null,
    val: number,
    depth: number
): TreeNode | null {
    // 特殊处理
    if (depth == 1) {
        return new TreeNode(val, root);
    }

    // depth >= 2 时
    const dfs = (node: TreeNode | null, cur: number) => {
        if (cur + 1 === depth) {
            node.left = new TreeNode(val, node.left, null);
            node.right = new TreeNode(val, null, node.right);
            return;
        }

        node.left && dfs(node.left, cur + 1);
        node.right && dfs(node.right, cur + 1);
    };

    dfs(root, 1);

    return root;
}

export function addOneRow(
    root: TreeNode | null,
    val: number,
    depth: number
): TreeNode | null {
    // 特殊处理
    if (depth == 1) {
        return new TreeNode(val, root);
    }

    // BSF
    const queue: TreeNode[] = [root];
    let cur = 1;

    while (queue.length) {
        const size = queue.length;
        for (let i = size; i > 0; i--) {
            const node = queue.shift();

            if (cur + 1 === depth) {
                node.left = new TreeNode(val, node.left, null);
                node.right = new TreeNode(val, null, node.right);
            }

            node.left && queue.push(node.left);
            node.right && queue.push(node.right);
        }
        cur++;
        if (cur === depth) {
            return root;
        }
    }

    return root;
}
